segmentation/src/OpticalFlowSegmentation.cpp

```diff
@@ -0,0 +1,493 @@
+                prev[1] + projLen * direction[1]);
+
+            nextVs[i] = Voxel(proj[0], proj[1], zIndex + 1);
+        }
+    }
+
+    // Return the updated vector of Voxel points
+    return nextVs;
+}
+
+auto OpticalFlowSegmentation::compute() -> PointSet
+{
+    // Reset progress
+    progressStarted();
+
+    // Duplicate the starting chain
+    auto currentVs = startingChain_;
+
+    // Update the user-defined boundary
+    bb_.setUpperBoundByIndex(2, endIndex_ + 1);
+
+    // Check that incoming points are all within bounds
+    if (std::any_of(begin(currentVs), end(currentVs), [this](auto v) {
+            return !bb_.isInBounds(v) || !vol_->isInBounds(v);
+        })) {
+        status_ = Status::ReturnedEarly;
+        progressComplete();
+        return create_final_pointset_({currentVs});
+    }
+
+    const fs::path outputDir("debugvis");
+    const auto wholeChainDir = outputDir / "whole_chain";
+    if (dumpVis_) {
+        fs::create_directory(outputDir);
+        fs::create_directory(wholeChainDir);
+    }
+
+    // Calculate the starting index
+    auto minZPoint = std::min_element(
+        begin(currentVs), end(currentVs),
+        [](const auto& a, const auto& b) { return a[2] < b[2]; });
+    auto startIndex = static_cast<int>(std::floor((*minZPoint)[2]));
+
+    if (endIndex_ <= startIndex) {
+        throw std::domain_error("end index <= start index");
+    }
+
+    // Collection to hold all positions
+    std::vector<std::vector<Voxel>> points;
+    points.reserve(
+        (endIndex_ - startIndex + 1) / static_cast<std::size_t>(stepSize_));
+    points.push_back(currentVs);
+
+    // Iterate over z-slices
+    std::size_t iteration{0};
+    auto stepSize = static_cast<int>(stepSize_);
+    const int padding = vol_->numSlices();
+    for (int zIndex = startIndex; zIndex <= endIndex_; zIndex += stepSize) {
+        // Update progress
+        progressUpdated(iteration++);
+
+        // Directory to dump vis
+        auto zStr = to_padded_string(zIndex, padding);
+        const fs::path zIdxDir = outputDir / zStr;
+
+        //////////////////////////////////////////////////////////
+        // 0. Resample current positions so they are evenly spaced
+        FittedCurve currentCurve(currentVs, zIndex);
+        currentVs = currentCurve.evenlySpacePoints();
+
+        // Dump entire curve for easy viewing
+        if (dumpVis_) {
+            const auto wholeChainPath = wholeChainDir / (zStr + "_chain.png");
+            WriteImage(
+                wholeChainPath,
+                draw_particle_on_slice_(currentCurve, zIndex, -1, true));
+        }
+
+        // Set up the maximum number of threads
+        auto maxThreads = std::thread::hardware_concurrency() - 1;
+        if (maxThreads_.has_value()) {
+            maxThreads = std::min(maxThreads, maxThreads_.value());
+        }
+
+        // Calculate the num threads we're going to use
+        const std::uint32_t minPointsPerThread{15};
```

Comment:
Tried to clean these calculations up so they're a little more readable. Hard task with all of the casting required, but what can you do?